src/storage-controller/src/collection_mgmt.rs

```diff
@@ -192,13 +347,476 @@ where
+                .map(|(row, diff)| Update {
+                    row: row.clone(),
+                    timestamp: self.current_upper.clone(),
+                    diff: diff.clone(),
+                })
+                .collect();
+
+            let now = T::from((self.now)());
+            let new_upper = std::cmp::max(
+                now,
+                TimestampManipulation::step_forward(&self.current_upper),
+            );
+
+            let request = vec![(
+                self.id,
+                updates_to_write,
+                self.current_upper.clone(),
+                new_upper.clone(),
+            )];
+
+            let append_result = match self.write_handle.compare_and_append(request.clone()).await {
+                // We got a response!
+                Ok(append_result) => append_result,
+                // Failed to receive which means the worker shutdown.
+                Err(_recv_error) => {
+                    // Sender hung up, this seems fine and can happen when
+                    // shutting down.
+                    notify_listeners(responders, || {
+                        Err(StorageError::ShuttingDown("PersistMonotonicWriteWorker"))
+                    });
+
+                    // End the task since we can no longer send writes to persist.
+                    return Err(());
+                }
+            };
+
+            match append_result {
+                // Everything was successful!
+                Ok(()) => {
+                    // Notify all of our listeners.
+                    notify_listeners(responders, || Ok(()));
+
+                    self.current_upper = new_upper;
+
+                    // Very important! This is empty at steady state, while
+                    // desired keeps an in-memory copy of desired state.
+                    self.to_write.clear();
+
+                    tracing::debug!(%self.id, "appended to incremental collection");
+
+                    // Break out of the retry loop so we can wait for more data.
+                    break 'append_retry;
+                }
+                // Failed to write to some collections,
+                Err(StorageError::InvalidUppers(failed_ids)) => {
+                    // Someone else wrote to the collection. We need to read
+                    // from persist and update to_write based on that and the
+                    // desired state.
+
+                    assert_eq!(
+                        failed_ids.len(),
+                        1,
+                        "received errors for more than one collection"
+                    );
+                    assert_eq!(
+                        failed_ids[0].id, self.id,
+                        "received errors for a different collection"
+                    );
+
+                    let actual_upper = failed_ids[0]
+                        .current_upper
+                        .as_option()
+                        .expect("cannot write to empty collection")
```

Comment:
Should we try to handle this gracefully? Or is there a reason why we can expect this to never happen?